sdk/clientcore/core/src/main/java/io/clientcore/core/models/CoreException.java

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package io.clientcore.core.models;
+
+import io.clientcore.core.implementation.http.RetryUtils;
+
+import java.io.UncheckedIOException;
+
+/**
+ * Base class for all exceptions thrown by the client libraries and top-level core components.
+ * <p>This class extends the {@code RuntimeException} class, which means that it is an unchecked exception.</p>
+ *
+ * Instances of this class or its subclasses are typically thrown in response to errors that occur when interacting
+ * with remote services. For example, if a network request to a remote service fails, a {@code CoreException} might
+ * be thrown.
+ * It also wraps local IO exceptions and deserialization errors.
+ * <p>
+ * Client libraries should implement their own exceptions that extend this class.
+ */
+public class CoreException extends RuntimeException {
+    /**
+     * Translates a {@link Throwable} into a {@link CoreException}.
+     *
+     * @param cause the {@link Throwable} to translate
+     * @return the {@link CoreException} that was created
+     */
+    public static CoreException from(Throwable cause) {
```

Comment:
We should also support creating a CoreException with just a message. For e.g. I should be able to `throw CoreException.from("something went wrong")` instead of creating another exception instance and then wrapping with CoreException. Currently, we have to use `null` to do this - `CoreException.from("something went wrong", null)`

Reply:
We can always add one, so far we don't have a case for it in core and I'd prefer someone who throws a brand new exception to at least consider a dedicated type for it